mass_mailing_custom_unsubscribe/controllers/main.py

```diff
@@ -0,0 +1,160 @@
+# License LGPL-3.0 or later (http://www.gnu.org/licenses/lgpl.html).
+
+from openerp import exceptions
+from openerp.http import local_redirect, request, route
+from openerp.addons.mass_mailing.controllers.main import MassMailController
+from .. import exceptions as _ex
+
+
+class CustomUnsuscribe(MassMailController):
+    def unsubscription_reason(self, mailing_id, email, res_id,
+                              qcontext_extra=None):
+        """Get the unsubscription reason form.
+
+        :param mail.mass_mailing mailing_id:
+            Mailing where the unsubscription is being processed.
+
+        :param str email:
+            Email to be unsubscribed.
+
+        :param int res_id:
+            ID of the unsubscriptor.
+
+        :param dict qcontext_extra:
+            Additional dictionary to pass to the view.
+        """
+        values = self.unsubscription_qcontext(mailing_id, email, res_id)
+        values.update(qcontext_extra or dict())
+        return request.website.render(
+            "mass_mailing_custom_unsubscribe.reason_form",
+            values)
+
+    def unsubscription_qcontext(self, mailing_id, email, res_id):
+        """Get rendering context for unsubscription form.
+
+        :param mail.mass_mailing mailing_id:
+            Mailing where the unsubscription is being processed.
+
+        :param str email:
+            Email to be unsubscribed.
+
+        :param int res_id:
+            ID of the unsubscriptor.
+        """
+        email_fname = origin_name = None
+        domain = [("id", "=", res_id)]
+        record_ids = request.env[mailing_id.mailing_model].sudo()
+
+        if "email_from" in record_ids._fields:
+            email_fname = "email_from"
+        elif "email" in record_ids._fields:
+            email_fname = "email"
+
+        if email_fname and email:
+            domain.append((email_fname, "ilike", email))
+        else:
+            # Trying to unsubscribe without email? Bad boy...
+            raise exceptions.AccessDenied()
+
+        if record_ids._name == "mail.mass_mailing.contact":
+            domain.append(
+                ("list_id", "in",
+                 [l.id for l in mailing_id.contact_list_ids]))
+
+        # Unsubscription targets
+        record_ids = record_ids.search(domain)
+
+        if not record_ids:
+            # Trying to unsubscribe with fake criteria? Bad boy...
+            raise exceptions.AccessDenied()
+
+        # Special data to load for some models
+        specials = {
+            "mail.mass_mailing.contact": {
+                "related": "list_id",
+                "origin": "display_name",
+            },
+            "crm.lead": {
+                "origin": "name",
+                "contact": "contact_name",
+            },
+            "hr.applicant": {
+                "related": "job_id",
+                "origin": "name",
+            },
+            # In case you install OCA's event_registration_mass_mailing
+            "event.registration": {
```

Comment:
The difference is you need to provide also the code for the matching and all the logic. Having the method, you only need to define the keys `related` and `origin` (and maybe `contact`), and reuse the same logic.